README.md
# Treeviz

This javascript module aims at providing an easy interface in order to represent tree diagrams on screen with the ability to handle dynamic data flows. The data format must be JSON.

[Treeviz Demo](https://codepen.io/pierrecapo/pen/MPbBdv)

## Installation

With npm : `npm install treeviz`

and then you can use it with : `import Treeviz from 'treeviz';`

Or download this repository and link the dist/index.js file in your page directly : `<script src="./dist/index.js><script>`

## Usage

This module comes with a bunch of configuration settings.

`var myTree = Treeviz.create({ // configuration here });`

Then you can easily pass data to the object, and it will generate the tree :

```
myTree.refresh(data);
myTree.refresh(data_update1);
myTree.refresh(data_update2);
```

The tree will be clever enough to updates only the part of the trees that have been added or removed in the dataset, and so it won't redraw the entire tree.

## API

The big part of the API is configuring the tree before passing data to it :

```js
Treeviz.create(config);
```

#### config.htmlD (String)

Mandatory. The HTML id tag on the page where the tree should be drawn.

#### config.nodeField (String)

Mandatory. The unique identifier field in the dataset representing the node

#### config.relationnalField (String)

Mandatory. This one is a bit harder to understand. There are 2 ways to provide data to the tree : or a flat dataset (typically an array of nodes), or a pre-built hierarchical dataset.
In case of flat dataset, usually the relationnal field between each node is the field representing the father of the node, linking it to the id of the field. (See example below).
In case of pre-built hierarchical dataset, the relationnal field is the field representing an array of children

#### config.areaWidth (String)

The width avalaible for drawing the tree. Default to "800px"

#### config.areaHeight (String)

The height of the drawn area. Default to "450px"

#### config.nodeSettings (Object)

An object defining the nodes. Default values are shown here

```js
{ width:  160, // width of the nodes
height:  100, // height of the nodes
depthDistance:  300, // distance between depth levels
colorField:  null, // Setting the node color. If null, nodes will be blue.
template:  null  /* HTML definition of the text inside the nodes.
                  {{ }} is the way to access the data field dynamically inside
                  the template, like : <div>{{my_field}}</div> */
}
```

#### config.linkSettings(Object)

An object defining the links. Default values are shown here

```js
{ colorField:  null, // Setting the link color. If null, links will be grey
widthField:  null /* field responsible of the width of links.
                  The module will automatically compute the max of this field
                  from every node, and will make a linear scale from 1px to 50px
                  with it */
}
```

#### config.horizontalLayout (String)

Toggle whether the tree should have an horizontal or vertical layout. Default to true (horizontal layout).

#### config.zoomBehavior (String)

Toggle whether pan or zoom are avalaible or not. Default to false (no zoom).

#### Treeviz.on (Event)

If you click on a node, the event will be triggered, and the node data will be passed as an argument to the callback function.

`Treeviz.on("nodeClick", function(d) { // do something });`

## Example

Assuming that you already have an HTML element on the pagh with the tag `id="a_tree"`

#### Flat data case :

```js
var flat_data_example = [
  { id: 0, node_name: "Root node", father: null, number: 10 },
  { id: 1, node_name: "Child A", father: 0, number: 5 },
  { id: 2, node_name: "Child B", father: 0, number: 3 },
  { id: 3, node_name: "Child C", father: 0, number: 2 },
];

var myTree = Treeviz.create({
  htmlID: "a_tree",
  nodeField: "id",
  flatData: true,
  relationnalField: "father",
  nodeSettings: {
    width: 160,
    height: 100,
    depthDistance: 300,
    template:
      "<div style='height:{{height}}px; width:{{width}}px;display:flex;justify-content:center;align-items:center;'><div>{{node_name}}</div></div>",
  },
});

myTree.refresh(flat_data_example);
```

#### Hierarchical data case :

```js
var hierarchical_data_example = {
  name: "Mom",
  qty: 10,
  children: [{ name: "Son A", qty: 3 }, { name: "Son B", qty: 7 }],
};

var myTree = Treeviz.create({
  htmlID: "a_tree",
  nodeField: "name",
  flatData: false,
  relationnalField: "children",
  nodeSettings: {
    width: 160,
    height: 100,
    depthDistance: 300,
    template:
      "<div style='height:{{height}}px; width:{{width}}px;display:flex;justify-content:center;align-items:center;'><div>{{name}}</div></div>",
  },
  linkSettings: {
    colorField: null,
    widthField: "qty",
  },
});

myTree.refresh(hierarchical_data_example);
```

## Contributing

1. Fork it!

2. Create your feature branch: `git checkout -b my-new-feature`

3. Commit your changes: `git commit -am 'Add some feature'`

4. Push to the branch: `git push origin my-new-feature`

5. Submit a pull request :D

## Credits

This module is based on d3 library, credit to all the contributors of this project.
Thanks also to the people from the Emitter project

## License

BSD


CHANGELOG.md
### 1.0.3

- Add a shape property to linkSettings. Allows to have orthogonal links or cubic Beziers links


src/index.ts
import Emitter from "component-emitter";
import * as d3 from "d3";
import { drawLinks } from "./draw-links";
import { initiliazeSVG } from "./initializeSVG";
import { prepareData } from "./prepare-data";
import { templateSystem } from "./template-system";

export const create = ({
  htmlID = "",
  nodeField = "",
  relationnalField = "",
  flatData = true,
  areaWidth = 800,
  areaHeight = 450,
  nodeSettings = {
    width: 160,
    height: 100,
    depthDistance: 300,
    colorField: "",
    template: "",
  },
  linkSettings = { colorField: "#A1887F", widthField: "" },
  horizontalLayout = true,
  zoomBehavior = false,
}) => {
  const settings = {
    htmlID,
    nodeField,
    relationnalField,
    flatData,
    areaWidth,
    areaHeight,
    nodeSettings,
    linkSettings,
    horizontalLayout,
    zoomBehavior,
  };

  const draw = (svg: any, source: any, treemap: any) => {
    let i = 0;
    const duration = 400;

    // Assigns the x and y position for the nodes
    const treeData = treemap(source);

    // Compute the new tree layout.
    const nodes = treeData.descendants();
    const links = treeData.descendants().slice(1);

    const linkScale = d3
      .scaleLinear()
      // @ts-ignore
      .domain([0, d3.max(nodes.map(el => el.data[linkSettings.widthField]))])
      .range([0, 50]);
    // Normalize for fixed-depth.
    nodes.forEach((d: any) => {
      d.y = d.depth * nodeSettings.depthDistance;
    });

    // ****************** Nodes section ***************************

    // Update the nodes...
    const node = svg.selectAll("g.node").data(nodes, (d: any) => {
      return d[nodeField] || (d[nodeField] = ++i);
    });

    // Enter any new modes at the parent's previous position.
    const nodeEnter = node
      .enter()
      .append("g")
      .attr("class", "node")
      .attr("transform", (d: any) => {
        if (horizontalLayout) {
          return typeof d.ancestors()[1] !== "undefined"
            ? "translate(" + d.ancestors()[1].y + "," + d.ancestors()[1].x + ")"
            : "translate(" + source.y0 + "," + source.x0 + ")";
        } else {
          return typeof d.ancestors()[1] !== "undefined"
            ? "translate(" + d.ancestors()[1].x + "," + d.ancestors()[1].y + ")"
            : "translate(" + source.x0 + "," + source.y0 + ")";
        }
      });

    // Add squares for the nodes
    nodeEnter
      .append("rect")
      .attr("class", "node")
      .attr("rx", 5)
      .attr("ry", 5)
      .attr("width", nodeSettings.width)
      .attr("height", nodeSettings.height)
      .style("fill", (d: any) => {
        return d[nodeSettings.colorField] || "#2196F3";
      })
      .style("cursor", "pointer")
      .on("click", (d: any) => {
        // @ts-ignore
        obj.emit("nodeClick", d);
      });

    nodeEnter
      .append("foreignObject")
      .attr("width", nodeSettings.width)
      .attr("height", nodeSettings.height)
      .style("pointer-events", "none")
      .html((d: any) =>
        templateSystem(
          Object.assign(
            {},
            d.data,
            { width: nodeSettings.width },
            { height: nodeSettings.height }
          ),
          nodeSettings.template
        )
      );

    // UPDATE
    const nodeUpdate = nodeEnter.merge(node);

    // Transition to the proper position for the node
    nodeUpdate
      .transition()
      .duration(duration)
      .attr("transform", (d: any) => {
        return horizontalLayout
          ? "translate(" + d.y + "," + d.x + ")"
          : "translate(" + d.x + "," + d.y + ")";
      });

    // Remove any exiting nodes
    const nodeExit = node
      .exit()
      .transition()
      .duration(duration)
      .attr("transform", (d: any) => {
        return horizontalLayout
          ? "translate(" + d.ancestors()[1].y + "," + d.ancestors()[1].x + ")"
          : "translate(" + d.ancestors()[1].x + "," + d.ancestors()[1].y + ")";
      })
      .remove();

    // On exit reduce the node circles size to 0
    nodeExit.select("circle").attr("r", 1e-6);

    // On exit reduce the opacity of text labels
    nodeExit.select("text").style("fill-opacity", 1e-6);

    // ****************** links section ***************************

    // Update the links...
    const link = svg.selectAll("path.link").data(links, (d: any) => {
      return d.id;
    });

    // Enter any new links at the parent's previous position.
    const linkEnter = link
      .enter()
      .insert("path", "g")
      .attr("class", "link")
      .attr("d", (d: any) => {
        const o = horizontalLayout
          ? { x: d.ancestors()[1].x, y: d.ancestors()[1].y }
          : { x: d.ancestors()[1].y, y: d.ancestors()[1].x };
        return drawLinks(o, o, settings);
      })
      .attr("fill", "none")
      .attr("stroke-width", (d: any) => {
        return linkScale(d.data[linkSettings.widthField]) || 10;
      })
      .attr("stroke", (d: any) => {
        return d[linkSettings.colorField] || "#2196F3";
      });

    // UPDATE
    const linkUpdate = linkEnter.merge(link);

    // Transition back to the parent element position
    linkUpdate
      .transition()
      .duration(duration)
      .attr("d", (d: any) => {
        return drawLinks(d, d.parent, settings);
      });

    // Remove any exiting links
    // @ts-ignore
    const linkExit = link
      .exit()
      .transition()
      .duration(duration)
      .attr("d", (d: any) => {
        const o = horizontalLayout
          ? { x: d.ancestors()[1].x, y: d.ancestors()[1].y }
          : { x: d.ancestors()[1].y, y: d.ancestors()[1].x };
        return drawLinks(o, o, drawLinks);
      })
      .remove();

    // Store the old positions for transition.
    nodes.forEach((d: any) => {
      d.x0 = d.x;
      d.y0 = d.y;
    });
  };

  function refresh(data: any): any {
    const [dataA, treemap] = prepareData(data, settings);
    draw(svg, dataA, treemap);
  }

  const obj = { refresh };
  Emitter(obj);

  const svg = initiliazeSVG(settings);
  return obj;
};
